aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -111,36 +111,63 @@ Tensor pinverse(const Tensor& self, double rcond) {
   return at::matmul(V.conj() * S_pseudoinv.unsqueeze(-2), U.transpose(-2, -1).conj());
 }
 
-static inline Tensor _matrix_rank_helper(const Tensor& self, bool symmetric) {
+Tensor linalg_matrix_rank(const Tensor& self, optional<double> tol, bool hermitian) {
+  // Matrices or batch of matrices are allowed
+  TORCH_CHECK(self.dim() >= 2, "linalg_matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ", self.sizes());
+
+  // NumPy doesn't take into account possible input with no elements and it errors on max not defined for this case
+  // Let's output 0 for this case, since that kind of matrices have zero number of non-zero rows, hence rank is 0.
+  if (self.numel() == 0) {
+    // matrix_rank assigns a scalar value for each matrix in the batch so
+    // result's shape is equal to self.shape[0:self.ndim-2]
+    // for single matrix result_shape = {}
+    auto result_shape = self.sizes().vec();
+    result_shape.pop_back();
+    result_shape.pop_back();
+    return at::zeros(result_shape, self.options().dtype(ScalarType::Long));
+  }
+
+  // We compute matrix rank as the number of singular or absolute eigen values above 'tol' threshold
   Tensor S;
-  if (!symmetric) {
+  if (!hermitian) {
     Tensor U, V;
+    // TODO: replace self.svd with linalg_svd
     std::tie(U, S, V) = self.svd(/*some=*/true, /*compute_uv=*/false);
   } else {
     Tensor eigvecs;
+    // TODO: replace self.symeig with linalg_eigvalsh
     std::tie(S, eigvecs) = self.symeig(/*eigenvectors=*/false);
     S = S.abs();
   }
-  return S;
+
+  if (tol.has_value()) {
+    double tol_value = tol.value();
+    return (S > tol_value).sum(/*dim=*/-1);
+  } else {
+    ScalarType real_dtype = toValueType(typeMetaToScalarType(self.dtype()));
+    double tol_value = _get_epsilon(real_dtype) * std::max(self.size(-1), self.size(-2));
+    Tensor max_S = std::get<0>(S.max(/*dim=*/-1));
+    return (S > max_S.mul_(tol_value).unsqueeze_(-1)).sum(/*dim=*/-1);
+  }
 }
 
-Tensor matrix_rank(const Tensor& self, double tol, bool symmetric) {
-  TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())) && self.dim() == 2,
-              "matrix_rank(", self.scalar_type(), "{", self.sizes(), "}): expected a 2D tensor "
-              "of floating types");
+// TODO: implement _out variant avoiding copy and using already allocated storage directly
+Tensor& linalg_matrix_rank_out(Tensor& result, const Tensor& self, optional<double> tol, bool hermitian) {
+  TORCH_CHECK(result.scalar_type() == ScalarType::Long,
+    "result dtype ", result.scalar_type(), " does not match the expected dtype ", ScalarType::Long);
 
-  Tensor S = _matrix_rank_helper(self, symmetric);
-  return (S > tol).sum();
+  Tensor result_tmp = at::linalg_matrix_rank(self, tol, hermitian);
+  at::native::resize_output(result, result_tmp.sizes());
+  result.copy_(result_tmp);
```

Comment:
same as with `cond` PR, I agree that the memory and perf savings by not copying results are minimal, but the general pattern for other pytorch ops is for `out` variants to directly use provided `result` tensor and not copy into it. It's fine to have it as a TODO in both cases though.

Reply:
Now the provided `result` is used directly here and `linalg_matrix_rank` is implemented via calling to `linalg_matrix_rank_out`. It's a bit easier here than in `cond` as we don't need to worry about autograd.